### PR TITLE
[Merged by Bors] - Add new validator API for voluntary exit

### PR DESCRIPTION
**Pull request by jimmygchen**
## Issue Addressed

Addresses #4117 

## Proposed Changes

See https://github.com/ethereum/keymanager-APIs/pull/58 for proposed API specification.

## TODO

- [x] ~~Add submission to BN~~ 
  - removed, see discussion in [keymanager API](https://github.com/ethereum/keymanager-APIs/pull/58)
- [x] ~~Add flag to allow voluntary exit via the API~~ 
  - no longer needed now the VC doesn't submit exit directly
- [x] ~~Additional verification / checks, e.g. if validator on same network as BN~~ 
  - to be done on client side
- [x] ~~Potentially wait for the message to propagate and return some exit information in the response~~ 
  - not required
- [x] Update http tests
- [x] ~~Update lighthouse book~~ 
  - not required if this endpoint makes it to the standard keymanager API

### CODE REVIEW

**Comment by jimmygchen** [comment]
Hi @AgeManning 

I've done an basic implementation based on the existing voluntary exit logic from account manager. I have a few questions:
- Is it necessary and possible here to check the validator is on the same network as the BN? I wonder if the BN already does some sort of checks. I think this may require passing the `Eth2NetworkConfig` to the `http-api::Context`, I'm checking if this is available in the VC.
  - **UPDATE**: Not required, this is already checked during VC startup
- I think it may be useful to wait for the message to propagate (one slot?) and return additional info in the response (e.g. `current_epoch`, `exit_epoch`, `withdrawal_epoch`). Alternatively, we could get the UI to query the validator a slot later - this might be easier. What do you think?
  -  **UPDATE**: best for the UI to query

**Comment by jimmygchen** [comment]
Looks like we may not need to submit the message to the BN with this API, we could return the signed voluntary exit object and leave it to the UI to publish to BN.

**Comment by AgeManning** [comment]
Looking good!

**Comment by jimmygchen** [comment]
The current implementation works, but I'd probably hold off further changes for now as we have an active discussion on this endpoint in the key manager API spec https://github.com/ethereum/keymanager-APIs/issues/57

**Comment by jimmygchen** [comment]
PR for keymanager-API spec: https://github.com/ethereum/keymanager-APIs/pull/58

File: validator_client/src/http_api/create_signed_voluntary_exit.rs
Comment (paulhauner) [inline validator_client/src/http_api/create_signed_voluntary_exit.rs]: Perhaps this should be a 404 "not found" rather than a 500 "internal server error"?

Comment (jimmygchen) [reply validator_client/src/http_api/create_signed_voluntary_exit.rs]: Good catch. Updated, thanks!

File: validator_client/src/http_api/create_signed_voluntary_exit.rs
Comment (paulhauner) [inline validator_client/src/http_api/create_signed_voluntary_exit.rs]: I was using this endpoint on Goerli on a VC with 5,000 validators on it (a pretty unreasonable amount of validators, I know). It turns out that it takes some time to obtain all the validator indices for all pubkeys when the VCs starts (on a fresh boot the VC only knows pubkeys but doesn't know validator indices). I was hitting this 404 and it took me a bit to figure out what was going on.

Here's a suggestion you're free to cherry-pick which would clarify things:

https://github.com/sigp/lighthouse/commit/b1a1cbfacc80d78556bd28aa2130de589ba53734

File: validator_client/src/http_api/create_signed_voluntary_exit.rs
Comment (paulhauner) [inline validator_client/src/http_api/create_signed_voluntary_exit.rs]:
```suggestion
    info!(
        log,
        "Signing voluntary exit";
        "validator" => pubkey_bytes.as_hex_string(),
        "epoch" => epoch
    );
```

Maybe it'll come in handy one day.

**Review by paulhauner — COMMENTED** [review]
Very nice, I used this a couple of days ago to produce a signed exit on Prater! It worked really well. I've just made a couple of very minor suggestions based on my experience.

Happy to merge this after they've been addressed!

File: validator_client/src/http_api/create_signed_voluntary_exit.rs
Comment (jimmygchen) [reply validator_client/src/http_api/create_signed_voluntary_exit.rs]: Hey @paulhauner, thanks for this, nice changes! 👍  I've cherry-picked your changes. I wasn't aware that the validator indices were fetched / available later - good to know!

**Review by paulhauner — APPROVED** [review]
Excellent, let's do it!

**Pull request closed without merging**
